package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -726,6 +772,670 @@ def __setstate__(self, state):
+
+    def _write_next_timestep(self, ts):
+        """Write coordinates and unitcell information to H5MD file.
+
+        Do not call this method directly; instead use
+        :meth:`write` because some essential setup is done
+        there before writing the first frame.
+
+        The first dimension of each dataset is extended by +1 and
+        then the data is written to the new slot.
+
+        Note
+        ----
+        Writing H5MD files with fancy trajectory slicing where the Timestep
+        does not increase monotonically such as ``u.trajectory[[2,1,0]]``
+        or ``u.trajectory[[0,1,2,0,1,2]]`` raises a :exc:`ValueError` as this
+        violates the rules of the step dataset in the H5MD standard.
+
+        """
+
+        i = self._counter
+
+        # H5MD step refers to the integration step at which the data were
+        # sampled, therefore ts.data['step'] is the most appropriate value
+        # to use. However, step is also necessary in H5MD to allow
+        # temporal matching of the data, so ts.frame is used as an alternative
+        self._step.resize(self._step.shape[0]+1, axis=0)
+        try:
+            self._step[i] = ts.data['step']
+        except(KeyError):
+            self._step[i] = ts.frame
+        if len(self._step) > 1 and self._step[i] < self._step[i-1]:
+            raise ValueError("The H5MD standard dictates that the step "
+                             "dataset must increase monotonically in value.")
+
+        # the dataset.resize() method should work with any chunk shape
+        self._time.resize(self._time.shape[0]+1, axis=0)
+        self._time[i] = ts.time
+
+        if 'edges' in self._traj['box']:
+            self._edges.resize(self._edges.shape[0]+1, axis=0)
+            self._edges.write_direct(ts.triclinic_dimensions,
+                                     dest_sel=np.s_[i, :])
+        # These datasets are not resized if n_frames was provided as an
+        # argument, as they were initialized with their full size.
+        if self._has['position']:
+            if self.n_frames is None:
+                self._pos.resize(self._pos.shape[0]+1, axis=0)
+            self._pos.write_direct(ts.positions, dest_sel=np.s_[i, :])
+        if self._has['velocity']:
+            if self.n_frames is None:
+                self._vel.resize(self._vel.shape[0]+1, axis=0)
+            self._vel.write_direct(ts.velocities, dest_sel=np.s_[i, :])
+        if self._has['force']:
+            if self.n_frames is None:
+                self._force.resize(self._force.shape[0]+1, axis=0)
+            self._force.write_direct(ts.forces, dest_sel=np.s_[i, :])
+
+        if self.data_keys:
+            for key in self.data_keys:
+                obs = self._obsv[f'{key}/value']
+                obs.resize(obs.shape[0]+1, axis=0)
+                obs[i] = ts.data[key]
+
+        if self.convert_units:
+            self._convert_units(i)
+
+        self._counter += 1
+
+    def _convert_units(self, i):
```

Comment:
Can you rename this to `_convert_dataset_with_units()` or something similarly expressive? Because for a second I though this was one of the standard `_convert_*` methods and wondered why you would want to convert the integer `i`.